src/dscanner/analysis/numbers.d

```diff
@@ -5,70 +5,82 @@
 
 module dscanner.analysis.numbers;
 
-import std.stdio;
-import std.regex;
-import dparse.ast;
-import dparse.lexer;
 import dscanner.analysis.base;
-import dscanner.analysis.helpers;
-import dsymbol.scope_ : Scope;
+import dmd.tokens : TOK;
+import std.conv;
+import std.regex;
+import std.stdio;
 
 /**
  * Checks for long and hard-to-read number literals
  */
-final class NumberStyleCheck : BaseAnalyzer
+extern (C++) class NumberStyleCheck(AST) : BaseAnalyzerDmd
 {
-public:
-	alias visit = BaseAnalyzer.visit;
-
+	alias visit = BaseAnalyzerDmd.visit;
 	mixin AnalyzerInfo!"number_style_check";
 
-	/**
-	 * Constructs the style checker with the given file name.
-	 */
-	this(BaseAnalyzerArguments args)
+	private enum KEY = "dscanner.style.number_literals";
+	private enum string MSG = "Use underscores to improve number constant readability.";
+
+	private auto badBinaryRegex = ctRegex!(`^0b[01]{9,}`);
+	private auto badDecimalRegex = ctRegex!(`^\d{5,}`);
+
+	extern (D) this(string fileName, bool skipTests = false)
 	{
-		super(args);
+		super(fileName, skipTests);
 	}
 
-	override void visit(const Token t)
+	override void visit(AST.IntegerExp intExpr)
 	{
-		import std.algorithm : startsWith;
+		import dscanner.utils : readFile;
+		import dmd.errorsink : ErrorSinkNull;
+		import dmd.globals : global;
+		import dmd.lexer : Lexer;
 
-		if (isNumberLiteral(t.type) && !t.text.startsWith("0x")
-				&& ((t.text.startsWith("0b") && !t.text.matchFirst(badBinaryRegex)
-					.empty) || !t.text.matchFirst(badDecimalRegex).empty))
-		{
-			addErrorMessage(t, "dscanner.style.number_literals",
-					"Use underscores to improve number constant readability.");
-		}
+		auto bytes = readFile(fileName) ~ '\0';
```

Comment:
Is it really necessary to reread the file? Don't we already have the source in memory?

Reply:
I don't think so. I'm thinking a proper solution for this would be a bigger refactor - a common lexer visitor base class, passing down a token stream to the visitors created from the inital d-scanner run, etc.